Remove engines and back relation mappers with veritas only

snusnu opened the pull request:
This removes the whole concept of engines as we had them before. Since we decided to back relational mappers with veritas only, they make no sense anymore. This PR introduces  `Repository` objects that can either be `InMemory` or `Persistent`. Persistent repositories keep track of gateway relations that provide access to relations using a given `adapter`. InMemory repositories simply keep track of `Veritas::Relation` instances.

Basically, this PR also introduces the possibility to run everything in memory, providing us with a way to run all our tests against the "reference implementation" (in addition to running them against any specific veritas adapter).

This also opens up further development / refactoring regarding `#insert`, `#update` and `#delete` which in turn should move us towards session integration. Nothing is stopping us from writing unit and integration specs for these operations, since we don't have to rely on (S)QL generation anymore to test our expectations.

I also think that it's worth to mention that once this PR is done, we made another tiny step towards making `dm-mapper` a more viable choice for developing apps even now. I expect most app development can (happily) accept to write tests against the in memory reference implementation, not caring about loosing data on shutdown. This in turn buys us some more time to finally implement full `#insert`, `#update` and `#delete` support for SQL generation.

snusnu committed:
Nuke all engines except for veritas

snusnu committed:
Move Engine::Veritas into {Engine, Relation::Graph::{Node, Edge}}

snusnu committed:
Remove commented code

snusnu committed:
Make the uri param required in Engine#initialize

snusnu committed:
Remove docs that no longer apply

snusnu committed:
Pass the already coerced uri down to DataObjects

snusnu committed:
Nuke pending flag from M:N self join specs < 1.9

snusnu committed:
Revert "Nuke pending flag from M:N self join specs < 1.9"

This reverts commit e321dc4670e427a64d7cc7e0b6f6ca34006a6b00.

snusnu committed:
Simplify Relationship::Iterator#each specs

This hopefully fixes the last remaining issue on
rbx-18

snusnu committed:
Don't allow rbx-18mode to fail on travis

snusnu committed:
Don't inject the engine into Relation::Graph#initialize

This was previously only used to determine the
#relation_node_class and #relation_edge_class to
use within the graph. Atm, this functionality is
not needed anymore.

I've changed Relation::Graph#initialize to support
injecting those two classes instead. However, I'm
not sure if that's really needed eventually. It's
not unlikely that this will be removed or changed
in some other not yet foreseeable way.

snusnu committed:
Inject the environment in Mapper#initialize

snusnu committed:
Keep the relation graph in the Environment instance

snusnu committed:
[WIP] Introduce DataMapper::Repository

snusnu committed:
Nuke Engine in favor of Environment#repositories

snusnu committed:
Move integration specs out of the veritas directory

snusnu committed:
Add IRC notification on successful builds

snusnu committed:
[ci skip] Update SpecHelper.draw_relation_graph

snusnu committed:
Push Mapper#environment down to Relation::Mapper

snusnu committed:
Push enumerable inclusion down to Relation::Mapper

snusnu committed:
Add LICENSE file

snusnu committed:
Add .yardopts file

snusnu committed:
Update gem dependencies

Temporarily depend on snusnu/veritas which has the
dependency on backports bumped to ~> 2.8.2

dkubb commented on lib/data_mapper/finalizer/base_relation_mappers_finalizer.rb:
```diff
@@ -53,8 +57,8 @@ def finalize_edges
-        mapper_registry << mapper.new(mapper.relations.node_for(mapper.gateway_relation))
+      def finalize_mapper(mapper, relation)
+        mapper_registry << mapper.new(environment, relation)
         mapper.freeze
```
It might be interesting to test what would happen if you deep freeze the mapper. In theory nothing extra should happen, but it can sometimes expose when you have immutable objects backed by mutable state; in general you want to eliminate this if you can.

snusnu committed:
Add .rvmrc file

snusnu committed:
Bump the abstract_type dependency to ~> 0.0.4

dkubb commented on .travis.yml:
```diff
@@ -17,9 +16,9 @@ rvm:
   - ruby-head
 matrix:
```
It looks like the whole `matrix` section can be removed now that it's passing with all rubies.

dkubb commented on Gemfile.devtools:
```diff
@@ -30,21 +30,19 @@ group :guard do
-  platforms :mri_18 do
-    gem 'rcov', '~> 1.0.0'
+  platforms :mri_19, :rbx do
+   gem 'mutant', '~> 0.2.17'
```
Indentation looks a bit off here.

dkubb commented on Rakefile:
```diff
@@ -7,22 +7,12 @@ Devtools.init
-  desc "Run specs on travis"
-  task :travis => %w(spec:unit spec:engines spec:isolation)
+  desc "Run all specs on travis"
+  task :travis => %w(spec:unit spec:integration spec:isolation)
```
Why have a special travis specific task? Why not do what the other DM2 related gems do and have a "ci" task that runs the specs, and metrics when it can?

snusnu replied:
@dkubb yeah, we should definitely unify that at some point soon.

dkubb commented on lib/data_mapper/environment.rb:
```diff
@@ -33,7 +33,7 @@ class Environment
     # @api private
     def initialize(registry = nil)
-      @engines = {}
+      @repositories = {}
```
I think it would be nicer to define this as `def initialize(registry = Undefined)` and then only do: reset(registry) unless `registry.equal?(Undefined)`. Technically it's more code, but it's more specific. I don't like the idea of code that behaves the same when called like:

``` ruby
DataMapper::Environment.new
DataMapper::Environment.new(nil)
```

It just seems if somewhere along the line code accidentally injects `nil` into the constructor the code will happily accept it, when it should throw an exception imho.

EDIT: IMHO any method that sets the default value to `nil` like this would be better off using the `Undefined` convention.

dkubb replied:
Or you could just change this method to be:

``` ruby
def initialize(registry = Mapper::Registry.new)
  @repositories = {}
  reset(registry)
end
```

If someone happens to pass in `nil` the method should explode imho.

dkubb commented on lib/data_mapper/environment.rb:
```diff
@@ -47,17 +47,26 @@ def initialize(registry = nil)
+    # @return [Repository]
+    #
+    # @api private
+    def repository(name)
```
I wish I had extracted the inheritable aliasing code from veritas, since this is an example where it could eliminate this method.

snusnu replied:
@dkubb it's probably never too late ;)

dkubb commented on lib/data_mapper/finalizer/base_relation_mappers_finalizer.rb:
```diff
@@ -16,18 +16,22 @@ def run
-        mapper.relations.new_node(name, relation, header)
+        relation_node = relations.build_node(name, relation, header)
+        relations.add_node(relation_node)
+        relation_node
```
This is a rather large method imho. I would suggest breaking it down into more well named methods rather than a single large block of code like it is.

dkubb commented on lib/data_mapper/environment.rb:
```diff
@@ -159,11 +166,16 @@ def finalize
 
+    protected
+
+    attr_reader :repositories
```
Since there is only a single protected method, I would be more inclined to define this attribute higher up in the class body and pass in the method name explicitly to protected, eg:

``` ruby
attr_reader :repositories
protected :repositories
```

dkubb commented on lib/data_mapper/relation/graph.rb:
```diff
@@ -266,7 +254,12 @@ def [](name)
-        self[relation.name.to_sym]
+        # TODO streamline this
+        if relation.respond_to?(:name)
+          self[relation.name.to_sym]
```
Is it necessary to call `#to_sym` on the relation name? You already do it in the `#[]` method, plus it should probably be part of the contract. I think unnecessary usage of `#to_*` should be removed from inside a system.. it's fine at the boundaries where you've got user input coming in, but once you get inside the system I think it's safe to assume a specific interface.

dkubb commented on lib/data_mapper/relation/graph/edge.rb:
```diff
@@ -112,16 +112,29 @@ def target_name
-        attr_reader :header
+        def join_relation(operation)
+          relation = source_relation.join(target_relation)
+          relation = relation.instance_eval(&operation) if operation
```
It looks like `operation` may always be defined, so it might just be nicer to do:

``` ruby
def join_relation(operation)
  source_relation.join(target_relation).instance_eval(&operation)
end
```

snusnu replied:
@dkubb actually, `operation` is not always defined, at least currently. Changing the code to what you suggested breaks a lot of specs. We could look at refactoring it at some point tho.

dkubb commented on lib/data_mapper/repository.rb:
```diff
@@ -0,0 +1,150 @@
+    # TODO make that smarter
+    def self.adapter(options)
+      parsed_uri = Addressable::URI.parse(options.fetch(:uri))
+      Veritas::Adapter::DataObjects.new(parsed_uri)
```
I assume this won't always be hard coded?

snusnu replied:
@dkubb definitely not.

dkubb commented on lib/data_mapper/repository.rb:
```diff
@@ -0,0 +1,150 @@
+  class Repository
+
+    def self.coerce(name, options = EMPTY_HASH)
+      if options.any?
```
Maybe I should make a `veritas-memory-adapter` gem to avoid special casing it?

dkubb replied:
Or you could make an adapter-like object that responds to `#gateway` that just passes through the relation?

dkubb commented on Gemfile:
```diff
@@ -4,29 +4,13 @@ gemspec
-    gem 'activerecord-jdbcpostgresql-adapter', '>= 1.2.0'
-  end
-end
+gem 'adamantium',            :git => 'https://github.com/dkubb/adamantium'
```
For consistency with the urls in the other Gemfile, should we append ".git" on them?  (Same with on https://github.com/datamapper/dm-mapper/pull/33/files#L3R23)

dkubb commented on lib/data_mapper/environment.rb:
```diff
@@ -47,17 +47,26 @@ def initialize(registry = nil)
     #
-    # @param [Hash] options
-    #   the options to use for instantiating the engine
+    # @param [Addressable::Uri, String] uri
```
This isn't how an option is documented in YARD. `@param` is for positional arguments. There should be one for `options` and then you can use the following for the individual options: http://rubydoc.info/docs/yard/file/docs/Tags.md#option

dkubb commented on lib/data_mapper/environment.rb:
```diff
@@ -159,11 +166,16 @@ def finalize
+      @mappers      = []
+      @registry     = registry || Mapper::Registry.new
+      @relations    = Relation::Graph.new
+      @finalized    = false
```
The spacing here looks a bit inconsistent.

dkubb commented on lib/data_mapper/relation/graph/node.rb:
```diff
@@ -140,8 +140,122 @@ def delete(key)
+        #
+        # @api public
+        def ordered
+          new(name, relation.sort_by(relation.header), header)
```
At the very least I would suggest using `#sort_by` in this class to cut down on duplication.

snusnu replied:
@dkubb mapper refactoring is next on our list once this PR is done.

dkubb commented on lib/data_mapper/repository.rb:
```diff
@@ -0,0 +1,150 @@
+
+    # TODO make that smarter
+    def self.adapter(options)
+      parsed_uri = Addressable::URI.parse(options.fetch(:uri))
```
IMHO if this isn't on the boundary of the system we should assume it is already an `Addressable::URI` instance and not parse it here.

snusnu replied:
@dkubb i expect these parts to see quite some refactoring still. they're basically just the simplest thing that works for now.

dkubb commented on lib/data_mapper/repository.rb:
```diff
@@ -0,0 +1,150 @@
+
+    private_class_method :adapter
+
+    include AbstractType
```
I would suggest having `extend` and `include` at the top, just after the class/module is declared. They affect behaviour so deeply and are as important as what class you inherit from, in the case of a class. I don't like it when this is buried in between some method declarations, it makes it really easy to miss.

dkubb commented on spec/db_setup.rb:
```diff
@@ -4,20 +4,18 @@
-    DM_ENV.setup(name, :uri => uri, :engine => engine_name.to_sym)
-  end
-end
+DM_ENV.setup(REPOSITORY, { :uri => URI })
```
These curly braces are unnecessary.

dkubb commented on spec/support/helper.rb:
```diff
@@ -1,34 +1,40 @@
+  #   the environment to use for accessing the relation graph
+  #
+  # @return [undefined]
+  def self.draw_relation_graph(env, file_name = 'graph.png')
```
Is this actually a method that supports the specs? Or should it be moved somewhere else?

snusnu committed:
Cleanup the travis configuration file

snusnu committed:
Push Environment#reset down to TestEnv

As #reset is only used from specs currently, there
is no need to expose it as a public method to
client code. Every user of dm-mapper will interact
with Environment directly, so we should only expose
what's really necessary.

Also, the mapper registry gets populated during
Environment#finalize, so a typical user would not
need to inject an instance. It seems that this
constructor injection was primarily done to "ease"
writing of specs (using #should_receive :/). IMHO, 
until a real usecase arises, it's better to just
simplify the public api of this class and thus not
support injecting a mapper registry. It would only
be confusing for users. Also, the behavior would
be kinda "undefined" anyway.

snusnu committed:
Make Environment#repositories protected explicitly

snusnu committed:
Split BaseRelationMappersFinalizer into smaller methods

snusnu committed:
Avoid useless #to_sym call in Relation::Graph#node_for

snusnu committed:
Consistently declare git dependencies in the Gemfile

snusnu committed:
Fix docs for Environment#setup

snusnu committed:
Move include directive to the top of the class body

snusnu committed:
Fix alignment of the private directive

snusnu committed:
Remove useless curly braces for hash definition

snusnu committed:
Move graphviz code to TestEnv#draw_relation_graph

snusnu commented:
@dkubb I picked all low hanging fruit you found. thx for looking over it! I've replied to all comments I haven't yet tackled with commits.

snusnu committed:
Bundle dkubb/veritas (backports dep is updated)

snusnu committed:
[ci skip] Rename TestEnv#{draw_relation_graph => draw}

dkubb commented on lib/data_mapper/finalizer/base_relation_mappers_finalizer.rb:
```diff
@@ -43,18 +41,31 @@ def finalize_relationships
 
+      def registered_node(mapper, name, header)
+        relation = registered_relation(name, mapper)
+        node = relations.build_node(name, relation, header)
```
It's interesting that we're passing in `relation` and `header` into this method. I know that `relation.header` is not exactly the same kind of object as in `header`; the latter more or less being the list of attributes in the object to map to (?). It seems as if these two ideas are relatively closely related within the system and should perhaps be contained together (I assume this is what the node's purpose is?).

My main question is if this information can be known earlier and/or should it be in a registry to avoid creating it over and over? It looks like that's what the code does, but I wanted to confirm since I don't see things at as high a level as you do yet.

snusnu committed:
[ci skip] Fix docs for Environment#setup

snusnu committed:
[ci skip] Move TestEnv#draw => Support::Graphviz#draw

snusnu committed:
Put InMemory + Persistent repositories in separate files

snusnu committed:
Close #34 Move support code below lib/data_mapper

snusnu committed:
Refactor Repository and Veritas::Adapter objects

* Collapse Repository::{InMemory, Persistent} into
  Repository only, which accepts a name + adapter
* Support registering veritas adapters by their
  URI scheme and implement the following dedicated
  adapters:
    Veritas::Adapter::Postgres
    Veritas::Adapter::InMemory
* Remove Environment#setup in favor of passing a
  repository config hash to Environment#initialize
  (typically used via Environment.coerce)
* Refactor spec_helpers to be self contained for
  a given spec run (unit/integration)

snusnu committed:
Remove duplicate OpenStruct monkey patch for specs

snusnu committed:
Remove unused rspec :focus metadata configuration

I actually remember that being kinda dangerous
from app development. If you happen to forget to
remove a :focus flag, only the focused tests will
be run.

During library development, I found myself simply
running the specs in question explicitly, using
something like:

  bundle exec rspec -cfs spec/file/path

snusnu committed:
Dry up spec helpers and use one configured DM_ENV

snusnu committed:
[ci skip] Correct (irrelevant) typo in database.yml

snusnu committed:
Use https://rubygems.org as the Gemfile source

Silences a bundler warning about :rubygems being
deprecated because of security concerns.

snusnu committed:
[ci skip] Add docs for Environment.coerce

snusnu committed:
[ci skip] Update README to reflect recent changes

snusnu committed:
Add specs for Environment.coerce

snusnu committed:
[ci skip] Nuke whitespace in specs

snusnu commented:
@solnic, @dkubb, @mbj i think i'm gonna merge this in soon, as the PR is already quite large. As a successor, I plan to add support for shared adapter specs in a subsequent PR. Fine with you guys?

mbj commented:
+1 from my side. Especially as I can use dm-mapper with this changes ;)

solnic commented:
Sure

snusnu committed:
Add class comment for Environment

snusnu committed:
Add docs for Veritas::Adapter::DataObjects#uri

snusnu committed:
Add docs for Veritas::Adapter::InMemory

snusnu committed:
Add module docs for Veritas::Adapter

snusnu committed:
Simplify db_setup (to be improved soon)

snusnu commented:
alright then ....
